### ISSUE TITLE
Validate primer combination allocation

### ISSUE THREAD

**Issue by charles-cowart** [jdereus/labman#415]
Output from shotgun library prep is similar to, but is not equal to the output from https://github.com/tanaes/metagenomics_pooling_notebook.git. Confirmed labcontrol.ucsd.edu and notebook both used test_output/iTru/new_iTru_combos_Dec2017.csv. Amanda and I are investigating.

**Comment by charles-cowart** [comment]
Amanda and I recreated the Echo pick-list files for the two plates that were created on 03-27-2019 and confirmed that the Labcontrol output from Rodolfo was from the first plate, with a starting combo index of 384, where previously we assumed it was 756.

Previously, we confirmed that the shotgun primer combo set IDs began at 1 in the labcontrol database, whereas in the ipython script, they begin at 0.

Both of the above conditions appear to be the cause of much of the initial discrepancy we saw in output between labcontrol and the script. Amanda is confirming that ordering of rows in the tab-delimited files is not a requirement, and Charlie is examining both outputs based on a starting index of 756.

An additional plate was created in Labcontrol.ucsd.edu that will need to be deleted, once I can confirm the proper way to do so.

**Comment by AmandaBirmingham** [comment]
Slight tweaks here: We confirmed that the "*_indices.txt" file from Rodolfo was from the first plate, which had a starting index of *1* (not 384) whereas we had previously assumed it was *768*.   Keep in mind @charles-cowart 's note about labcontrol combo ids starting at 1 while the combo indexes in the lab's ipython notebooks begin at 0 :)  

Re deleting the additional plate in LabControl: let's just "discard" it (which can be done in the GUI interface), rather than trying to delete it, which would be (a) a real pain to clean up all the wells and compositions and processes associated with it and (b) contravening LabControl's axioms (e.g., "deletion = bad" :)

**Comment by AmandaBirmingham** [comment]
>Amanda is confirming that ordering of rows in the tab-delimited files is not a requirement
Confirmed.  Per meeting with wet-lab on 10/092018:

library prep echo picklist:
file name: date_nameofcompressedplate_indices.txt
**doesn't need to be sorted** [emphasis added]
index combo: don't need to add
sample: ok to have study id prefix
source plate name: replace all whitespace with an underscore
                # replace whitespace with underscore
                dna_plate_name = re.sub('\s+', '_', dna_plate_name)

Note that this means we should NOT expect the outputs of LabControl to be EXACTLY THE SAME as the outputs of the current process for the same data .... we can't just diff them and say "if there's a difference then LabControl is wrong".  We have to check if they contain the same contents (even if the order is not the same) and if they do, then LabControl is performing as designed.

**Comment by charles-cowart** [comment]
Gotcha. I made a Jupyter notebook last night, comparing the results from each. A few tweeks and I'll send you a copy.

**Comment by charles-cowart** [comment]
Just to recap a few points from @AmandaBirmingham 's email to everyone:

We reset the library_prep_shotgun_composition primary key to a value above 21505, so future library_prep_shotgun_composition records that go in should have no more “this key has already been used” errors (Caused by incorrectly updating library_prep_shotgun_composition's primary key to 21120.)

We reset the current_combo_index field in the record for the current shotgun primer set in labman.shotgun_primer_set to 21120.

We ran a test to ensure that we could create a new shotgun library prep plate through the LabControl GUI without errors and that the indices assigned started at the expected value.  After doing this, we repeated step 2 to put the current_combo_index back to what it should be for the plate the user wants to create.

We anticipate the results in Labcontrol will match the output from the iPython Notebook, if one does not consider that the rows need to be sorted.
